streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals.suppress;
+
+import org.apache.kafka.streams.KeyValue;
+
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+class InMemoryTimeOrderedKeyValueBuffer<K, V> implements TimeOrderedKeyValueBuffer<K, V> {
+    private long memBufferSize = 0;
+    private final Map<K, TimeKey<K>> index = new HashMap<>();
+    private final TreeMap<TimeKey<K>, SizedValue<V>> sortedMap = new TreeMap<>();
+
+    private static class SizedValue<V> {
+        private final V v;
+        private final long size;
+
+        private SizedValue(final V v, final long size) {
+            this.v = v;
+            this.size = size;
+        }
+
+        public V value() {
+            return v;
+        }
+
+        public long size() {
+            return size;
+        }
+
+        @Override
+        public String toString() {
+            return "SizedValue{v=" + v + ", size=" + size + '}';
+        }
+    }
+
+    private class InnerIterator implements Iterator<KeyValue<TimeKey<K>, V>> {
+        private final Iterator<Map.Entry<TimeKey<K>, SizedValue<V>>> delegate = sortedMap.entrySet().iterator();
+        private Map.Entry<TimeKey<K>, SizedValue<V>> last;
+
+        @Override
+        public boolean hasNext() {
+            return delegate.hasNext();
+        }
+
+        @Override
+        public KeyValue<TimeKey<K>, V> next() {
+            final Map.Entry<TimeKey<K>, SizedValue<V>> next = delegate.next();
+            this.last = next;
+            return new KeyValue<>(next.getKey(), next.getValue().value());
+        }
+
+        @Override
+        public void remove() {
+            delegate.remove();
+            index.remove(last.getKey().key());
+            memBufferSize = memBufferSize - last.getValue().size();
+        }
```

Comment:
Should we check for `last == null` and set `last = null` ?

Reply:
Good idea, setting it to null after I use it will make it available for gc.

I can guard against null also, but fwiw, I'm not sure how that situation could arise. It's an `IllegalStateException` to invoke  `delegate.remove` without an intervening call `delegate.next`. Or to call it before `next`.
`delegate.next` could return null, but in that case, we'd get an exception in line 69... which I should check for there.

Reply:
Ok, I just confirmed that `TreeMap#entrySet().iterator().next()` can never return `null`, but we could theoretically store a null value in the map, which could still throw an NPE on this line. I'll guarded against it.